From f65078016123fab206fac64deee214f5bdc0ab0f Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Tue, 30 Dec 2025 11:37:22 +0800
Subject: [PATCH 22/71] video: rockchip: rga3: fix page fault caused by IOMMU
 prefetch

1. IOMMU prefetch requires configuring min/max limits based on the prefetch
size.
2. Fix misuse of ALIGN/ALIGN_DOWN macros.

Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
Change-Id: Id9d27a3b1a644de1d43976eebefa8259f1ef08cb
(cherry picked from commit e78e66240ba657c4de38ae3766f87f253dba3984)
---
 .../rockchip/rga3/include/rga2_reg_info.h     |  7 +++---
 drivers/video/rockchip/rga3/rga2_reg_info.c   | 25 ++++++++++++++-----
 2 files changed, 23 insertions(+), 9 deletions(-)

diff --git a/drivers/video/rockchip/rga3/include/rga2_reg_info.h b/drivers/video/rockchip/rga3/include/rga2_reg_info.h
index 5f4238b84eb6a..f19a2421b888e 100644
--- a/drivers/video/rockchip/rga3/include/rga2_reg_info.h
+++ b/drivers/video/rockchip/rga3/include/rga2_reg_info.h
@@ -534,13 +534,14 @@
 #define RGA2_VSP_BICUBIC_LIMIT				1996
 #define RGA2_BILINEAR_PREC				12
 
+#define RGA2_IOMMU_PREFETCH_SIZE			0x20000	//128k
 #define RGA2_IOMMU_PREFETCH_SHIFT			16
-#define RGA2_IOMMU_PREFETCH_MASK			0xffff
+#define RGA2_IOMMU_PREFETCH_MASK			(RGA2_IOMMU_PREFETCH_SIZE - 1)
 #define RGA2_IOMMU_PREFETCH_ALIGN(x) \
 	(((x) + RGA2_IOMMU_PREFETCH_MASK) & ~RGA2_IOMMU_PREFETCH_MASK)
 #define RGA2_IOMMU_PREFETCH_ALIGN_DOWN(x)		((x) & ~RGA2_IOMMU_PREFETCH_MASK)
-#define RGA2_IOMMU_PREFETCH_THRESHOLD_MIN		(0x1U << RGA2_IOMMU_PREFETCH_SHIFT)
-#define RGA2_IOMMU_PREFETCH_THRESHOLD_MAX		(0xffffU << RGA2_IOMMU_PREFETCH_SHIFT)
+#define RGA2_IOMMU_PREFETCH_THRESHOLD_MIN		RGA2_IOMMU_PREFETCH_SIZE
+#define RGA2_IOMMU_PREFETCH_THRESHOLD_MAX		(0x100000000ULL - RGA2_IOMMU_PREFETCH_SIZE)
 
 #define RGA2_MAX_RD_OUTSTANDING				0x3f
 
diff --git a/drivers/video/rockchip/rga3/rga2_reg_info.c b/drivers/video/rockchip/rga3/rga2_reg_info.c
index 577374fb2635b..236dfa4e7d054 100644
--- a/drivers/video/rockchip/rga3/rga2_reg_info.c
+++ b/drivers/video/rockchip/rga3/rga2_reg_info.c
@@ -297,8 +297,21 @@ static void RGA2_set_mode_ctrl(u8 *base, struct rga2_req *msg)
 		reg = ((reg & (~m_RGA2_MODE_CTRL_SW_TILE4x4_OUT_EN)) |
 		       (s_RGA2_MODE_CTRL_SW_TILE4x4_OUT_EN(1)));
 
-	reg = ((reg & (~m_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE)) |
-	       (s_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE(0))); /* 128k */
+	switch (RGA2_IOMMU_PREFETCH_SIZE) {
+	case 0x20000: /* 128k */
+		reg = ((reg & (~m_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE)) |
+		       (s_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE(0)));
+		break;
+	case 0x30000: /* 256k */
+		reg = ((reg & (~m_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE)) |
+		       (s_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE(1)));
+		break;
+	case 0x10000: /* 64k */
+	default:
+		reg = ((reg & (~m_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE)) |
+		       (s_RGA2_MODE_CTRL_SW_TABLE_PRE_FETCH_MODE(2)));
+		break;
+	}
 
 	if (msg->src.rd_mode == RGA_RKFBC_MODE || msg->src.rd_mode == RGA_AFBC32x8_MODE)
 		reg = ((reg & (~m_RGA2_MODE_CTRL_SW_FBC_IN_EN)) |
@@ -1683,18 +1696,18 @@ static void RGA2_set_reg_dst_info(u8 *base, struct rga2_req *msg)
 	if (rot_90_flag == 1) {
 		if (y_mirr == 1) {
 			msg->iommu_prefetch.y_threshold = y_lt_addr >> 16 ?
-				RGA2_IOMMU_PREFETCH_ALIGN_DOWN(y_lt_addr) :
+				RGA2_IOMMU_PREFETCH_ALIGN(y_lt_addr) :
 				RGA2_IOMMU_PREFETCH_THRESHOLD_MIN;
 			msg->iommu_prefetch.uv_threshold = u_lt_addr >> 16 ?
-				RGA2_IOMMU_PREFETCH_ALIGN_DOWN(u_lt_addr) :
+				RGA2_IOMMU_PREFETCH_ALIGN(u_lt_addr) :
 				RGA2_IOMMU_PREFETCH_THRESHOLD_MIN;
 		} else {
 			msg->iommu_prefetch.y_threshold = (y_rd_addr >> 16) == 0xffff ?
 				RGA2_IOMMU_PREFETCH_THRESHOLD_MAX :
-				RGA2_IOMMU_PREFETCH_ALIGN(y_rd_addr);
+				RGA2_IOMMU_PREFETCH_ALIGN_DOWN(y_rd_addr);
 			msg->iommu_prefetch.uv_threshold = (u_rd_addr >> 16) == 0xffff ?
 				RGA2_IOMMU_PREFETCH_THRESHOLD_MAX :
-				RGA2_IOMMU_PREFETCH_ALIGN(u_rd_addr);
+				RGA2_IOMMU_PREFETCH_ALIGN_DOWN(u_rd_addr);
 		}
 	}
 }
